### PR TITLE
refactor: change retry loop functions to accept enum in lieu of bool

### LINKED ISSUE #4975: Use an enum for "is_idempotent" parameter in retry loops
Our retry loops receive a `bool is_idempotent` parameter to determine if the operation is retryable or not.  Maybe an enum with values `kNotIdempotent` and `kIdempotent` would be easier to read.

### PR DESCRIPTION
**Pull request by scotthart**
fixes #4975

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977)

### CODE REVIEW

File: google/cloud/internal/retry_policy.h [hunk 1]
```diff
@@ -24,6 +24,9 @@ namespace google {
 namespace cloud {
 inline namespace GOOGLE_CLOUD_CPP_NS {
 namespace internal {
+
+enum class IsIdempotent { kIdempotent, kNonIdempotent };
```
Comment (devbww) [inline google/cloud/internal/retry_policy.h:28]: The "Is" part seems to re-imply the binary/booly-ness.

Might I suggest something more like `Idempotency"?

File: google/cloud/bigtable/internal/unary_client_utils.h [hunk 2]
```diff
@@ -136,7 +138,8 @@ struct UnaryClientUtils {
       bigtable::MetadataUpdatePolicy const& metadata_update_policy,
       MemberFunction function,
       typename Signature<MemberFunction>::RequestType const& request,
-      char const* error_message, grpc::Status& status, bool retry_on_failure) {
+      char const* error_message, grpc::Status& status,
+      google::cloud::internal::Idempotency retry_on_failure) {
```
Comment (devbww) [inline google/cloud/bigtable/internal/unary_client_utils.h:142]: Nit: `retry_on_failure` doesn't seem like the best name now.

**Review by devbww — APPROVED** [review]
This could certainly be done in a possible 2nd phase, but it seems that the remaining boolean `is_idempotent()` and `IsIdempotent()` functions would now be better returning an `Idempotency`.

**Review by coryan — CHANGES_REQUESTED** [review]
A few more changes below.

Candidly, I am not sure this is an improvement, but I am too comfortable with `bool` maybe.  I put it as a topic for today, it will probably take two minutes because I expect everybody but me to be in favor of the enum.

> Reviewed 27 of 27 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-:-MGIruBB42KwvrA2zEgi:bcq336g)* status: all files reviewed, 7 unresolved discussions (waiting on @devbww and @scotthart)

___
*[google/cloud/bigtable/table.cc, line 124 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MGIpfxA53FeIdJw1JvQ:-MGIpfxA53FeIdJw1JvR:bkm416y) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/bigtable/table.cc#L124)):*
> ```cpp
>   // it once and use a constant value for the loop.
>   auto idempotent_mutation_policy = clone_idempotent_mutation_policy();
>   auto const idempotency =
> ```

nit: maybe break into two lines?

```cc
auto const is_idempotent = /* as before */
auto const idempotency = is_idempotent ? Idempotency::kIdempotent : Idempotency::kNonIdempotent;
```
___
*[google/cloud/bigtable/internal/async_retry_op.h, line 47 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MGIqA5LC-lVDoyiwdD3:-MGIqA5LC-lVDoyiwdD4:b-s784mc) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/bigtable/internal/async_retry_op.h#L47)):*
> ```objc
>       : idempotency_(idempotency) {}
> 
>   bool is_idempotent() const {
> ```

maybe add another accessor `idempotency() const { return idempotency_; }` so this is easier to use?  Right now you are doing the equivalent of: `idempontency_ == Idempotency::kIdempotent : Idempotency::kIdempotent ? kNonIdempotent;` at each call site.
___
*[google/cloud/bigtable/internal/bulk_mutator.cc, line 60 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MGIq_065uANZq5Mcsab:-MGIq_065uANZq5Mcsac:b8rg8tq) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/bigtable/internal/bulk_mutator.cc#L60)):*
> ```cpp
>                            return idempotent_policy.is_idempotent(m);
>                          })
>                  ? Idempotency::kIdempotent
> ```

As before, maybe break into two lines.
___
*[google/cloud/bigtable/internal/unary_client_utils.h, line 142 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MFmVjuZZ-C2XLvzT5vs-r2-142:-MGIqx9k7IyA-LwjGiCi:be9qtgq) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/bigtable/internal/unary_client_utils.h#L142)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: `retry_on_failure` doesn't seem like the best name now.
</blockquote></details>

Or maybe this should remain a `bool` because the semantics are whether to retry or not.
___
*[google/cloud/internal/retry_loop.h, line 40 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MGIrAreAkd7po8IzXRz:-MGIrAreAkd7po8IzXS-:bj0kbsk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/internal/retry_loop.h#L40)):*
> ```objc
>  * @param backoff_policy controls how the loop backsoff from a recoverable
>  *     failure.
>  * @param idempotency if `Idempotencu::kNonIdempotent`, the operation is not
> ```

typo: `enc**u**`
___
*[google/cloud/pubsub/subscription_admin_connection.cc, line 145 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MGIrPMC-_eBFxFNAKLc:-MGIrPMC-_eBFxFNAKLd:bewfz2v) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/pubsub/subscription_admin_connection.cc#L145)):*
> ```cpp
>   StatusOr<google::pubsub::v1::Snapshot> CreateSnapshot(
>       CreateSnapshotParams p) override {
>     auto const is_idempotent = p.request.name().empty()
> ```

s/is_idempotent/idempotency/ maybe?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977)* status: 19 of 27 files reviewed, 6 unresolved discussions (waiting on @coryan and @devbww)

___
*[google/cloud/bigtable/internal/async_retry_op.h, line 47 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MGIqA5LC-lVDoyiwdD3:-MGKbXic5c2n14T00l4z:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/bigtable/internal/async_retry_op.h#L47)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

maybe add another accessor `idempotency() const { return idempotency_; }` so this is easier to use?  Right now you are doing the equivalent of: `idempontency_ == Idempotency::kIdempotent : Idempotency::kIdempotent ? kNonIdempotent;` at each call site.
</blockquote></details>

Done.
___
*[google/cloud/bigtable/internal/bulk_mutator.cc, line 60 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MGIq_065uANZq5Mcsab:-MGKTy4S204pyIrx3S6d:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/bigtable/internal/bulk_mutator.cc#L60)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

As before, maybe break into two lines.
</blockquote></details>

Done.
___
*[google/cloud/bigtable/internal/unary_client_utils.h, line 142 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MFmVjuZZ-C2XLvzT5vs-r2-142:-MGKUGfgD8UsWEMsojGk:b-msuw8j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/bigtable/internal/unary_client_utils.h#L142)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Or maybe this should remain a `bool` because the semantics are whether to retry or not.
</blockquote></details>

Per our discussion on bool parameters, I kept this type as the enum and updated the parameter name.
___
*[google/cloud/internal/retry_loop.h, line 40 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MGIrAreAkd7po8IzXRz:-MGKUmHg5zM9qVYFMOFd:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/internal/retry_loop.h#L40)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

typo: `enc**u**`
</blockquote></details>

Done.
___
*[google/cloud/pubsub/subscription_admin_connection.cc, line 145 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977#-MGIrPMC-_eBFxFNAKLc:-MGKV4_5CplMHOOX_DeZ:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f69c9c86b4caa13ae25b1d7995d3b8dbc8290b53/google/cloud/pubsub/subscription_admin_connection.cc#L145)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

s/is_idempotent/idempotency/ maybe?
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4977)* status: all files reviewed, 1 unresolved discussion (waiting on @devbww)

**Pull request merged**